edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/modules/DockerModule.cs

```diff
@@ -133,6 +139,18 @@ protected override void Load(ContainerBuilder builder)
                 .As<Task<IEnvironmentProvider>>()
                 .SingleInstance();
 
+            // Task<IBackupSource>
+            builder.Register(
+                async c =>
+                {
+                    var serde = c.Resolve<ISerde<DeploymentConfigInfo>>();
+                    IEncryptionProvider encryptionProvider = await c.Resolve<Task<IEncryptionProvider>>();
```

Comment:
This will work correctly because the await is last, but autofac has issues resolving after an await. This is because is uses the current thread context to store some information. To avoid this becoming an issue, we always store the resolved value as a task and await it in the constructor. This way it won't be an issue in the future if another parameter is added to the constructor. 
```suggestion
                    Task<IEncryptionProvider> encryptionProvider = c.Resolve<Task<IEncryptionProvider>>();
                    IBackupSource backupSource = new FileBackup(this.backupConfigFilePath, serde, await encryptionProvider);
```

Reply:
Thanks, I didn't remember what the specific issue with resolve/await was, one of the reasons I asked for your review.

Reply:
@lfitchett I also curious about such an issue. Could you share a link to issue?

Reply:
https://autofaccn.readthedocs.io/en/latest/advanced/concurrency.html